standard/expressions.md

```diff
@@ -1824,21 +1824,39 @@ A *null_conditional_projection_initializer* is a restriction of *null_conditiona
 
 ### 12.8.9 Null-forgiving expressions
 
-This operator sets the null state ([§8.9.5](types.md#895-nullabilities-and-null-states)) of the operand to “not null”.
+A null-forgiving expression’s value, type, classification ([§12.2](expressions.md#122-expression-classifications))
+and safe-context ([§16.4.12](structs.md#16412-safe-context-constraint)) is the value, type, classification and safe-context of its *primary_expression*.
 
 ```ANTLR
 null_forgiving_expression
-    : primary_no_array_creation_expression suppression
+    : primary_expression null_forgiving_operator
     ;
 
-suppression
+null_forgiving_operator
     : '!'
     ;
 ```
 
-This operator has no runtime effect; it evaluates to the result of its operand, and that result retains that operand’s classification.
+*Note*: The postfix null-forgiving and prefix logical negation operators ([§12.9.4](expressions.md#1294-logical-negation-operator)), while represented by the same lexical token (`!`), are distinct. Only the latter may be overriden ([§15.10](classes.md#1510-operators)), the definition of the null-forgiving operator is fixed. *end note*
+
+<!-- [Placeholder] 12.8.9 A -->
+<!-- [Placeholder] 12.8.9 A End -->
```

Comment:
Here I'd prefer the first alternative. Not specifying that would limit later grammar.

Reply:
@BillWagner mentions the known limitation argument.

Now I might be playing the devil’s advocate here, or maybe not 😉, but why does it not apply to *prefix* `!` or to operators such as `~`?

One could also ask why it doesn’t apply to negation as well, the following crazy code is valid (***please*** don't write it in real code!):

> ```
> int m = 0;
> int n = - - - --m;
> ```

which sets `m` to `-1` and `n` to `1`.

Given examples like this the apparent harmlessness of `x!!` stands out – it might earn a warning, but syntactic or semantic rules to prevent it?

Of course having the semantic rule as proposed here is also pretty harmless.

Reply:
+1 for the first alternative.